src/app.js

```diff
@@ -1,23 +1,146 @@
+require('dotenv').config()
 const express = require('express')
+const bodyParser = require('body-parser')
+const mongoose = require('mongoose')
+const bcrypt = require('bcryptjs')
+
+const saltRounds = 10
+
 const app = express()
-const port = 3000
+const port = 5000
+
+mongoose.connect(process.env.AUTH_DB, { useNewUrlParser: true, useUnifiedTopology: true })
 
+app.use(express.static('public'))
+app.set('view engine', 'ejs')
+app.use(bodyParser.urlencoded({
+  extended: true
+}))
+
+// Home Page
 app.get('/', (req, res) => {
-  res.send('Hello world!')
+  res.render('home')
 })
 
-app.post('/', (req, res) => {
-  res.send('Got a POST request!')
+// User Registration
+const userSchema = new mongoose.Schema({
+  email: String,
+  password: String,
+  key: String
 })
 
-app.put('/user', (req, res) => {
-  res.send('Got a PUT request at /user')
+const Model = mongoose.model
+const User = new Model('User', userSchema)
+
+const GenKey = () => {
+  // create a base-36 string that is always 30 chars long a-z0-9
+  // 'an0qrr5i9u0q4km27hv2hue3ywx3uu'
+  return [...Array(30)]
+    .map((e) => ((Math.random() * 36) | 0).toString(36))
+    .join('')
+}
+
+app.get('/register', (req, res) => {
+  res.render('register')
 })
 
-app.delete('/user', (req, res) => {
-  res.send('Got a DELETE request at /user')
+// Process of creating a user
+app.post('/register', (req, res) => {
+  // TODO: Implement duplicate-key checking logic
+  const apiKey = GenKey()
+
+  // Encrypt password
+  bcrypt.hash(req.body.password, saltRounds, (err, hash) => {
+    if (err) {
+      console.log(err)
+    } else {
+      const newUser = new User({
+        email: req.body.username,
+        password: hash,
+        key: apiKey
+      })
+
+      newUser.save((err) => {
+        if (err) {
+          console.log(err)
+        } else {
+          res.render('account', {
+            key: apiKey
+          })
+        }
+      })
+    }
+  })
 })
 
-app.listen(port, () => {
-  console.log(`Example app listening at http://localhost:${port}`)
+// Login Process
+app.get('/login', (req, res) => {
+  res.render('login')
 })
+
+app.post('/login', (req, res) => {
+  const userName = req.body.username
+  const password = req.body.password
+
+  User.findOne({ email: userName }, (err, foundUser) => {
+    if (err) {
+      console.log(err)
+    } else {
+      if (foundUser) {
+        bcrypt.compare(password, foundUser.password, (err, result) => {
+          if (!err) {
+            if (result === true) {
+              res.render('account', {
+                key: foundUser.key
+              })
+            } else {
+              console.log('Either username and/or password details are incorrect!')
+            }
+          } else {
+            console.log(err)
+          }
+        })
+      }
+    }
+  })
+})
+
+// Protected Routes to get information about UW Courses
+app.get('/courses', (req, res) => {
+  const params = req.query
+
+  const mockCourseInfo = {
```

Comment:
leftover test data?

Reply:
Ah, so basically what req.query does is get the query parameters off of a url/endpoint, so it's vital for making sure someone entered in an API Key. The mockCourseInfo is there temporarily until we have the schema finalized and I can start making models for the relational database

Reply:
Makes sense :0

Reply:
This works, another way of getting a 'it works!' response is replying back with a simple 'hitpoint successfully reached, data not available yet' message